### PR TITLE
Buffer system refactor

### PR DESCRIPTION
**Pull request by YiweiHan**

### CODE REVIEW

File: src/custom_message_sender/package.xml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="yiweih@todo.todo">yiweih</maintainer>
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Comment (hjed) [inline src/custom_message_sender/package.xml]: nit: please add the license here

File: src/custom_message_sender/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        for (uint32_t s = 0;s < sequences.size();++s) {
+            if (i < sequences[s].size()) {
+                flat.push_back(sequences[s][i]);
+            }
+        }
+    }
+    return flat;
+}
+
+int main(int argc, char *argv[]) {
```
Comment (hjed) [inline src/custom_message_sender/src/main.cpp:33]: this is a test file right? Can you add a comment somewhere to that effect

File: src/ros_can_nodes/include/CANBuffers.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Date Started: 20/10/2017
+ * Original Author: Simon Ireland
```
Comment (hjed) [inline src/ros_can_nodes/include/CANBuffers.hpp:3]: is this right?

File: src/ros_can_nodes/include/CANBuffers.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * Editors:
+ * ROS Node Name:
+ * ROS Package: ros_can_nodes
+ * Purpose:
```
Comment (hjed) [inline src/ros_can_nodes/include/CANBuffers.hpp]: please add this

File: src/ros_can_nodes/src/CANMsgRouter.cpp [hunk 5]
```diff
@@ -31,7 +36,7 @@ int main(int argc, char **argv) {
 
 void CANMsgRouter::init() {
     // TODO: either fail on bad open_port OR have reconnect policy
-    int err = CANHelpers::open_port("can0");
+    int err = CANHelpers::open_port("vcan0");
```
Comment (hjed) [inline src/ros_can_nodes/src/CANMsgRouter.cpp]: you need to change this back - maybe we should make it a ros param so its easy to change for testing

**Review by hjed — CHANGES_REQUESTED** [review]
Logic looks good - main thing is changing the device name back to the none virtual one. 
Have you tested this on a PCB? Let me know if you want me to do that

File: src/ros_can_nodes/src/CANMsgRouter.cpp (hunk 5 above)
Comment (YiweiHan) [reply src/ros_can_nodes/src/CANMsgRouter.cpp]: 👍

File: src/ros_can_nodes/include/CANBuffers.hpp (hunk 3 above)
Comment (YiweiHan) [reply src/ros_can_nodes/include/CANBuffers.hpp:3]: technically yes. this was renamed from TopicBuffers, but i pretty much overwrote the whole thing with this refactor

File: src/ros_can_nodes/src/CANMsgRouter.cpp (hunk 5 above)
Comment (YiweiHan) [reply src/ros_can_nodes/src/CANMsgRouter.cpp]: gone for a argc/argv approach for now

**Review by hjed — APPROVED** [review]

**Comment by hjed** [comment]
Approved assuming its working on hardware?

**Comment by YiweiHan** [comment]
yeah got it to work after the latest commit here. the embedded promise filter needs the sequence number to be 0 in the response message, so i used the inserter to insert seq 0, but the inserter had a bug where it wasnt setting the bits to 0 before setting the new value. once both were fixed everything worked

**Pull request merged**